lib/core/src/Cardano/Wallet/Primitive/Types/UTxOIndex/Internal.hs

```diff
@@ -343,6 +363,37 @@ selectRandom u selectionFilter =
         Any -> Map.keysSet $ utxo u
         WithAdaOnly -> entriesWithAdaOnly u
         WithAsset a -> entriesWithAsset a u
+        WithAssetOnly a -> entriesWithAssetOnly a u
+
+-- | Selects an entry at random from the index according to the given filters.
+--
+-- This function traverses the specified list of filters in descending order of
+-- priority, from left to right.
+--
+-- When considering a particular filter:
+--
+--    - if the function is able to select a UTxO entry that matches, it
+--      terminates with that entry and an updated index with the entry removed.
+--
+--    - if the function is not able to select a UTxO entry that matches, it
+--      traverses to the next filter available.
+--
+-- This function returns 'Nothing' if (and only if) it traverses the entire
+-- list of filters without successfully selecting a UTxO entry.
+--
+selectRandomWithPriority
+    :: MonadRandom m
+    => UTxOIndex
+    -> NonEmpty SelectionFilter
+    -- ^ A list of selection filters to be traversed in descending order of
+    -- priority, from left to right.
+    -> m (Maybe ((TxIn, TxOut), UTxOIndex))
+selectRandomWithPriority u filters =
+    selectRandom u (NE.head filters) >>= \case
+        Just en -> pure $ Just en
+        Nothing -> case NE.nonEmpty $ NE.tail filters of
+            Just fs -> selectRandomWithPriority u fs
+            Nothing -> pure Nothing
```

Comment:
Debatable to have switched to `NonEmpty SelectionFilter` after all :upside_down_face:

Reply:
> Debatable to have switched to `NonEmpty SelectionFilter` after all upside_down_face

I did indeed hesitate before doing that. On balance, since there doesn't seem to be a use case for calling this function with an empty list of filters, I thought it would be prudent to make it clear in the type signature.

Though I was a bit surprised to find out that there doesn't seem to be a `tailSafe :: NonEmpty a -> Maybe (NonEmpty a)` defined anywhere. (Though maybe I haven't looked hard enough.)